docs/rfcs/xxxx-river.md

```diff
@@ -0,0 +1,399 @@
+which compartmentalize the different configurable pieces of the agent, allowing
+users to more easily understand and debug configuration issues. Grafana Agent
+Flow was purposefully scoped broadly to allow for exploring many different
+component-based approaches for prototyping the experimental feature.
+
+The current implementation strategy focuses around expressions: settings for
+components can be derived from expressions which can reference and mutate the
+outputs of other components. Values can refer to arbitrary Go values like
+interfaces or channels, enabling component developers to easily allow users to
+construct data pipelines using Go APIs without requiring knowledge of the
+underlying implementation.
+
+The initial expressions prototype used [HCL][], which initially fit Flow's
+needs during early prototyping. However, the growing dependency on passing
+around arbitrary Go values to build pipelines started to conflict with the
+limitations of HCL, making HCL increasingly insufficient for Flow's specific
+use case.
+
+We examined alternatives to HCL such as YAML, CUE, Jsonnet, Lua, and Go itself.
+Eventually, we determined that the way we use arbitrary Go values in
+expressions for constructing pipelines is a new use case warranting a
+custom-built language.
+
+This document proposes River, an HCL-inspired declarative expressions-based
+language for continuous runtime evaluation. The decision to propose a new
+language is not taken lightly, and is seen as the last resort. As such, much of
+this proposal will focus on the rationale leading to this choice.
+
+## Goals
+
+* Minimize learning curve as much as possible to reduce friction
+* Make it easy for developers to consume for Flow components
+* Expose error messages in an easily understandable and actionable way
+* Natively support using Go values of any type
+* Natively support passing around and invoking real Go functions
+
+The language design will be scoped as small as possible, and new features will
+only be added over time as they are determined to be strictly necessary for
+Flow.
+
+## Non-Goals
+
+We are not aiming to create a general purpose configuration language. While it
+would be possible for River to eventually be used in different contexts by
+different projects, the primary goal today is specifically targeting Grafana
+Agent Flow.
+
+We will not provide a full specification for River here, only lightly
+describing it to allow implementation details to change over time.
+
+## Rationale
+
+### Why an expression language? Why not YAML?
+
+The entire rationale for creating a new language depends on the rationale that
+expressions provide a useful amount of capabilities to users. Expressions
+enable users to manipulate values to meet their own use cases in ways that
+otherwise would require dedicated feature work, such as:
+
+* Allowing users to merge metadata together from distinct sources when adding
```

Comment:
Metadata feels loose here. Is there a more concrete example?

Reply:
An example could be the [tenant](https://grafana.com/docs/loki/latest/clients/promtail/stages/tenant/) pipeline stage from Loki; a metrics component that would allow multitenancy based on specific rules?

Reply:
Added the example I was thinking of here:

> * Allowing users to merge metadata together from distinct sources when adding
  labels to metrics, such as merging labels from discovered Kubernetes
  namespaces with discovered Kubernetes pods.